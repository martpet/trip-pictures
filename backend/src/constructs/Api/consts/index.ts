export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
