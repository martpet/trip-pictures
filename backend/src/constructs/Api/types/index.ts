export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
