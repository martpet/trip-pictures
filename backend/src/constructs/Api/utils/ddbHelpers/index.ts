export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
