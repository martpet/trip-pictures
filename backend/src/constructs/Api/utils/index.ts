export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
