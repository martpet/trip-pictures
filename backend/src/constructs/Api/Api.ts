import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
