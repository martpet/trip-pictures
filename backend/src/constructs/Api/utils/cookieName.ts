import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
