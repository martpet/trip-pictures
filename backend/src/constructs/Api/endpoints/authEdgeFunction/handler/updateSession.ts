import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';

import { region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';
import { createDynamoUpdateExpression } from '~/utils';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type UpdateSessionProps = {
  sessionId: string;
  idToken: string;
};

export const updateSession = ({ sessionId, idToken }: UpdateSessionProps) => {
  const { tableName, partitionKey } = sessionsTableOptions;

  const data: Partial<SessionsTableItem> = {
    idToken,
  };

  const updateCommand = new UpdateCommand({
    TableName: tableName,
    Key: { [partitionKey.name]: sessionId },
    ...createDynamoUpdateExpression(data),
  });

  return ddbDocClient.send(updateCommand);
};
