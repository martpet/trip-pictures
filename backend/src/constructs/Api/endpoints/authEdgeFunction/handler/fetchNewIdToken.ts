import fetch from 'node-fetch';

import { FetchNewIdTokenResponse } from '~/constructs/Api/types';
import { authPaths } from '~/consts';

import { updateSession } from './updateSession';

type FetchNewIdTokenProps = {
  refreshToken: string;
  sessionId: string;
  clientId: string;
};

export const fetchNewIdToken = async ({
  refreshToken,
  sessionId,
  clientId,
}: FetchNewIdTokenProps) => {
  const { authDomain } = globalAuthEdgeFunctionProps;

  console.info('Fetching new id token...');

  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'refresh_token',
      client_id: clientId,
      refresh_token: refreshToken,
    }),
  });

  const data = (await response.json()) as FetchNewIdTokenResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  await updateSession({ sessionId, idToken: data.id_token });

  console.info('New id token fetched and saved');

  return data.id_token;
};
