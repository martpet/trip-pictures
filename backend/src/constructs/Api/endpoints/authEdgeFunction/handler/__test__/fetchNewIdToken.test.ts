import fetch, { Response } from 'node-fetch';

import { itCalls, itRejects, itResolves } from '~/constructs/Api/utils';

import { fetchNewIdToken } from '../fetchNewIdToken';
import { updateSession } from '../updateSession';

vi.mock('node-fetch');
vi.mock('../updateSession');

const args = [
  {
    refreshToken: 'dummyRefreshToken',
    sessionId: 'dummySessionId',
    clientId: 'dummyClientId',
  },
] as Parameters<typeof fetchNewIdToken>;

vi.mocked(fetch).mockResolvedValue({
  json: () => Promise.resolve({ id_token: 'dummyIdToken' }),
} as Response);

global.globalAuthEdgeFunctionProps = {
  authDomain: 'dummyAuthDomain',
} as typeof globalAuthEdgeFunctionProps;

describe('fetchNewIdToken', () => {
  itResolves(fetchNewIdToken, args);
  itCalls(fetch, fetchNewIdToken, args);
  itCalls(updateSession, fetchNewIdToken, args);

  describe('when "fetch" response contains an "error" prop', () => {
    beforeEach(() => {
      const error = { error: 'dummyErrorMessage' };
      const response = { json: () => Promise.resolve(error) } as Response;
      vi.mocked(fetch).mockResolvedValueOnce(response);
    });
    itRejects(fetchNewIdToken, args);
  });
});
