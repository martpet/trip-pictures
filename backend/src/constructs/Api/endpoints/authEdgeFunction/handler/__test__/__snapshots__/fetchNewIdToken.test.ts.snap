// Vitest Snapshot v1

exports[`fetchNewIdToken > calls "fetch" with correct args 1`] = `
[
  [
    "https://dummyAuthDomain/oauth2/token",
    {
      "body": URLSearchParams {
        Symbol(query): [
          "grant_type",
          "refresh_token",
          "client_id",
          "dummyClientId",
          "refresh_token",
          "dummyRefreshToken",
        ],
        Symbol(context): null,
      },
      "headers": {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchNewIdToken > calls "updateSession" with correct args 1`] = `
[
  [
    {
      "idToken": "dummyIdToken",
      "sessionId": "dummySessionId",
    },
  ],
]
`;

exports[`fetchNewIdToken > resolves with a correct value 1`] = `"dummyIdToken"`;

exports[`fetchNewIdToken > when "fetch" response contains an "error" prop > rejects with a correct value 1`] = `[Error: dummyErrorMessage]`;
